### PR TITLE
feat(flags): Improve UX around rollout, add blast radius

### PR DESCRIPTION
**Pull request by neilkakkar**
The problem: People sometimes accidentally roll out flags to everyone when they don't mean to. This gives a concrete look at how many users would this roll out to.

As a follow up, I'll add a pop confirmation when the total rollout is above X% of users.

See: https://github.com/PostHog/posthog/issues/12410

## Problem

~~Still a couple of bugs to fix & clean up to do~~, but in a place to solicit feedback on the UX.

Tech constraints:

Getting an accurate measurement of how many users will be affected, efficiently is surprisingly hard: (you can't just add all conditions in a query since the rollout percentages can be different, which means getting all personIDs and de-duplicating and cleverly applying the % condition so it's stable).

I felt it wasn't worth doing this, as even if we are accurate, it's going to be way too slow.

Further, every new condition added would mean recomputing all conditions, and it's pretty hard to then attribute %s to a specific condition.

So, I opted to compute blast radius independently per condition. This means we'll be pretty wrong for an overall number *if* some conditions overlap, but this case is small enough that this seemed fine to me.

On the plus side, this means we can compute things pretty fast, are approximately correct, and new conditions don't affect counts for existing conditions.

---

Here's how it looks like:  https://posthog.slack.com/archives/C034XD440RK/p1668697858514709

image

## Changes

1. [x] Estimates rollout blast radius per condition
2. [x] Partial filters not allowed anymore

TODO:
3. [ ] Confirmation dialog when new conditions increase total rollout by X%.
4. [ ] 

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

File: posthog/api/feature_flag.py [hunk 1]
```diff
@@ -339,6 +344,23 @@ def evaluation_reasons(self, request: request.Request, **kwargs):
 
         return Response(flags_with_evaluation_reasons)
 
+    @action(methods=["POST"], detail=False)
+    def user_blast_radius(self, request: request.Request, **kwargs):
```
Comment (EDsCODE) [inline posthog/api/feature_flag.py:348]: might make sense to cache this to some degree like we do with insights. if someone is adding and removing conditions in rapid succession we might dispatch too many queries

File: frontend/src/lib/constants.tsx [hunk 2]
```diff
@@ -132,6 +132,8 @@ export const FEATURE_FLAGS = {
     VARIANT_OVERRIDES: 'variant-overrides', // owner: @neilkakkar
     ONBOARDING_V2_EXPERIMENT: 'onboarding-v2-experiment', // owner: #team-growth
     RECORDING_AUTOPLAY: 'recording-autoplay', // owner: #team-session-recordings
+    SIGNUP_FORM_EXPERIMENT: 'signup-form-experiment', // owner: #team-growth
```
Comment (EDsCODE) [inline frontend/src/lib/constants.tsx:135]: why is this here

**Comment by liyiy** [comment]
1. indefinite loading for all users match
Screen Shot 2022-11-29 at 12 08 50 PM

2. > This means we'll be pretty wrong for an overall number if some conditions overlap, but this case is small enough that this seemed fine to me.

Can you elaborate more on this? I'm testing around and quite a number of cases are showing the wrong value for me, I think this might be a net negative vs neutral experience if it's not really working

3. Love the blocking of partial filters here! It's probably worth it to clean it up so that the other valid filters aren't showing an error though
Screen Shot 2022-11-29 at 12 12 59 PM

**Comment by liyiy** [comment]
Can you explain a little more on what we're trying to improve here? What do users gain from this and how does it improve their experience using feature flags? 🤗

**Comment by neilkakkar** [comment]
>  I'm testing around and quite a number of cases are showing the wrong value for me

Elaborate which cases are wrong? Any condition match should be extremely accurate.

It's the *total sum* that might not be accurate, but I don't show that anywhere on the UI, so it shouldn't be a big deal.


> Love the blocking of partial filters here! It's probably worth it to clean it up so that the other valid filters aren't showing an error though

D'oh, missed this, thanks!

> indefinite loading for all users match
checking 👀 

---

> Can you explain a little more on what we're trying to improve here?

The problem: People sometimes accidentally roll out flags to everyone when they don't mean to. This gives a concrete look at how many users would this roll out to.

As a follow up, I'll add a pop confirmation when the total rollout is above X% of users.

See: https://github.com/PostHog/posthog/issues/12410

File: frontend/src/lib/constants.tsx (hunk 2 above)
Comment (neilkakkar) [reply frontend/src/lib/constants.tsx:135]: Because I merged incorrectly, it seems like

File: posthog/api/feature_flag.py (hunk 1 above)
Comment (neilkakkar) [reply posthog/api/feature_flag.py:348]: Makes sense. I cached the total, but didn't consider cats rolling on the keyboard 😂 - will add

**Comment by liyiy** [comment]
> Elaborate which cases are wrong? Any condition match should be extremely accurate.

Hmmm I think this is working actually, I think it's because there were scenarios where I'd expect it to show users match but it doesn't (and is probably accurate).

**Review by liyiy — APPROVED** [review]
Other than the two bugs, looking good!

**Comment by neilkakkar** [comment]
The always loading on new flag (point. no 1 you raised) seems to be an issue with the router API, which I'll fix separately.

Every time you click new flag after the first page load works 😅 .

Have addressed no 3.
